### PR TITLE
Adds pre-commit hooks for linting and formatting

### PR DESCRIPTION
**Pull request by stumpylog**
This pull request add a `.pre-commit-config.yaml`, the configuration file for https://pre-commit.com/

At the moment, I have individually run the hooks to configure them a little bit, for things like ignoring certain files or running only on certain file types, etc.

Questions for discussion:
- Are the listed hooks beneficial?
- Are there further file type, directory, etc restrictions to place?
- Are there [more hooks](https://pre-commit.com/hooks.html) that might be good to add?
- Where is a good place to document installing the hooks?

Then once the hook list is agreed upon, a separate pull request (or several) can start cleaning up source files based on the linters and formatters.

To run the hooks manually and view the diffs this would produce:
`pre-commit run --all-files`

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      - id: trailing-whitespace
+        exclude_types:
+          - svg
+      - id: check-case-conflict
+      - id: detect-private-key
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: "v2.5.1"
+    hooks:
+      - id: prettier
+        types_or:
```
Comment (stumpylog) [inline .pre-commit-config.yaml:32]: This could also include YAML, Markdown, CSS, HTML formatting, if desired

**Comment by qcasey** [comment]
Thanks for putting this together! What's the status here, ready for review?

**Comment by stumpylog** [comment]
Sorry, my rebasing was probably spamming you.

From my point of view, this is ready basically, with the main outstanding question would be about where or if to document the installation of the hooks. 

Otherwise, it's a good base of relevant hooks to the project and it's not hard to add more if wanted.

I'll verify tomorrow the Pipfile, etc are fully up to date, but that doesn't affect the actual hooks.

**Comment by qcasey** [comment]
No worries at all. I'll test the hooks tomorrow.

Would also love some input from the front/back teams before merging this.

**Comment by stumpylog** [comment]
Ok, I've done a `pipenv lock` and confirmed that's all up to date.

I also ran the hooks against a copy of this branch, so the diff is visible here: [example formatting](https://github.com/stumpylog/paperless-ngx/compare/pre-commit-hooks...stumpylog:example-pre-commit-format).  If/as hooks change here, I'll try to keep that branch updated and in sync.

**Comment by qcasey** [comment]
I haven't heard of pre-commit before. This is great, very clean way to keep multiple formats enforced. @shamoon this [touches nearly all the js](https://github.com/stumpylog/paperless-ngx/compare/pre-commit-hooks...stumpylog:example-pre-commit-format), would like to hear your thoughts on the formatting.

### Docs

The hooks should absolutely be documented.

1. I believe this can replace the entire style section of [contributing.rst](https://github.com/paperless-ngx/paperless-ngx/blob/dev/docs/contributing.rst#how-to-get-your-changes-rolled-into-paperless). That doc is due for removal anyway, so I'm glad this simplifies it.
2. It should also be mentioned in both "development" sections of [extending.rst](https://github.com/paperless-ngx/paperless-ngx/blob/dev/docs/extending.rst#back-end-development). 

### Process

I'm wondering if this should be tested for/enforced in CI? Let's say someone opens a PR adding a scanner or fixing a typo from GitHub web editor, pre-commit won't run automatically. That'd get merged, and some future PR will run the format creating unrelated diffs. Is that true, if so how should we reckon?

**Comment by shamoon** [comment]
Wow yea this is epic. So after this is committed it will run it in future commits but we're also talking about running it on the existing code? How exactly does one do that? If we do choose to I'd like to do it myself so that it can be checked, etc. 

Also how can I tweak the [js] code style settings? Some of the choices I disagree with and are definitely accepted [js] standards like leaving off semicolons. Once we can tweak the settings it would also reduce the amount of changes.

Agree with other stuff qcasey said

**Comment by stumpylog** [comment]
I'll take a shot at updating the documentation as soon as I can.

About running it, by default `pre-commit` runs only on the files in a commit.  So if a contributor changes only a Python file, the hooks will run only for that file.  So a Python change won't affect the Javascript files with unrelated formatting changes.  The giant diff is only because I told it to run again every file in the repo.  My sort of idea to reduce the pain of reviewing the formatting changes would be breaking it into several pull requests, each targeting just a set of hooks, ie a Python formatting only pull.

The idea of getting it integrated into CI somehow is great.  It's not something I've done before, but it seems likely to be possible.  These hooks could basically replace the `codestyle` and `codeformatting` in that case.

@shamoon All the Javascript, Typescript formatting comes from the prettier hook.  Kind of like black for Python, it calls itself "opinionated".  [It can be configured some though](https://prettier.io/docs/en/options.html).  I noticed for example semicolons can be *not* added.

**Comment by shamoon** [comment]
Thanks for explanation. Yes I think we would break up the changes by file type or whatever. I'll look at prettier

**Comment by qcasey** [comment]
> About running it, by default pre-commit runs only on the files in a commit. So if a contributor changes only a Python file, the hooks will run only for that file. So a Python change won't affect the Javascript files with unrelated formatting changes.

Oh wow that's great. We don't need to worry about CI then IMO, that can be later on if still wanted/necessary.

> All the Javascript, Typescript formatting comes from the prettier hook.

Is there a less restrictive JS hook we can use, so shamoon can tweak to their liking?

**Comment by stumpylog** [comment]
> Is there a less restrictive JS hook we can use, so shamoon can tweak to their liking

There certainly might be, but a quick look through the "official" hooks on the website doesn't reveal anything right away.  If it can't be configured enough, it could be disabled for the Javascript and Typescript files, while being left on for other types (like Markdown and/or ReStructuredText, etc).  Or removed entirely for now.

Meanwhile, I took a stab at updating the docs to include installing the hooks and a brief blurb about how to make the hooks happy if they refuse a commit.

**Comment by qcasey** [comment]
Thanks a lot! Docs look good, some warnings but I expect the lint later will fix.

I'll leave the JS/TS decision up to shamoon, but I can say the other formatters are great and would love to have them.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          - javascript
+          - ts
+          - markdown
+        exclude: "(^Pipfile\\.lock$)"
```
Comment (shamoon) [inline .pre-commit-config.yaml:37]:
```suggestion
        exclude: "(^Pipfile\\.lock$)"
        args:
          - "--no-semi"
          - "--single-quote"
```

**Comment by shamoon** [comment]
Ok for the JS/TS part I think I'm good with the two options I suggested. And yea, in general this is awesome.

Is there anything that prevents someone from _not_ using this? I guess not, so they can make commits and they'd be fixed by CI afterward?

**Comment by qcasey** [comment]
Cool! 

> Is there anything that prevents someone from not using this? I guess not, so they can make commits and they'd be fixed by CI afterward?

In my example, someone contributing from the github web UI would not run this. In that case it can be corrected by either:

1. Formatted when someone else (who is using pre-commit) commits to that specific file later.
2. Formatted in CI (not implemented in this PR). This will be trickier like stumptylog said.

**Comment by shamoon** [comment]
Cool.

Yea this is really awesome. I use the Github desktop app too and you get this popup that warns you something was changed (yes I wish it was prettier, but Im ok with that 🤪).

Now I want to use this on all my projects!

Once its merged I will take care of js/ts files, I will leave the rest to you guys (or whoever). That sound ok?

**Review by shamoon — APPROVED** [review]

**Review by qcasey — APPROVED** [review]
Sounds good to me. Thank you @stumpylog!

File: .pre-commit-config.yaml (hunk 2 above)
Comment (stumpylog) [reply .pre-commit-config.yaml:40]: I've added those two options for the prettier formating hook.  [The diff](https://github.com/stumpylog/paperless-ngx/compare/pre-commit-hooks...stumpylog:example-pre-commit-format) is now updated as well.

**Comment by shamoon** [comment]
Yep, thanks =)

**Comment by shamoon** [comment]
Hey guys, I was thinking about handling PRs that havent run the pre-commits, two thoughts:

1. We could create a PR template which includes instructions or reminder (and / or checkbox) for having run them. Also maybe in general we should have a PR template?
2. Found this: https://github.com/marketplace/actions/prettier-action

In particular # 2 has a 'dry' mode that wont change code but at least can alert the author / us that the Prettier formatting failed (again thats the pre-commit hook we're using for js/ts/md).

Thoughts?

**Comment by stumpylog** [comment]
Yeah, that's a situation I haven't needed to consider before.  I've used the hooks with just myself before and I always approve.

1. I like the idea of a simple reminder to run the hooks in the pull request.  A template sounds great.
2. The pre-commit docs have a [section about using it in ci](https://pre-commit.com/#usage-in-continuous-integration).  There's also [pre-commit.ci](https://pre-commit.ci/).  

I'll look into these options. I'm a little less of a fan of the CI application, since I don't think it would be able to run against pull requests from forks and seems focused on changing the files, instead of just reporting a pass/fail status.  But adding a few actions to run the formatting checks, the flake8, maybe others seems doable.

**Pull request merged**